src/error.rs

```diff
@@ -0,0 +1,96 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! A collection of all possible errors encountered in PLONK.
+
+/// Defines all possible errors that can be encountered in PLONK.
+#[derive(core::fmt::Debug)]
+pub enum Error {
+    /// This error occurs when an error triggers on any of the fft module
+    /// functions.
+    InvalidEvalDomainSize {
+        /// Log size of the group
+        log_size_of_group: u32,
+        /// Two adacity generated
+        adacity: u32,
+    },
+    /// This error occurs when a proof verification fails.
+    ProofVerificationError,
+    /// This error occurs when the circuit is not provided with all of the
+    /// required inputs.
+    CircuitInputsNotFound,
+    /// This error occurs when we want to verify a Proof but the pi_constructor
+    /// attribute is uninitialized.
+    UninitializedPIGenerator,
+    /// PublicInput serialization error
+    InvalidPublicInputBytes,
+    /// This error occurs when an error triggers during the preprocessing
+    /// stage.
+    MismatchedPolyLen,
+    /// This error occurs when the Prover structure already contains a
+    /// preprocessed circuit inside, but you call preprocess again.
+    CircuitAlreadyPreprocessed,
+    /// This error occurs when the user tries to create PublicParameters
+    /// and supplies the max degree as zero.
+    DegreeIsZero,
+    /// This error occurs when the user tries to trim PublicParameters
+    /// to a degree that is larger than the maximum degree.
+    TruncatedDegreeTooLarge,
+    /// This error occurs when the user tries to trim PublicParameters
+    /// down to a degree that is zero.
+    TruncatedDegreeIsZero,
+    /// This error occurs when the user tries to commit to a polynomial whose degree is larger than
+    /// the supported degree for that proving key.
+    PolynomialDegreeTooLarge,
+    /// This error occurs when the user tries to commit to a polynomial whose degree is zero.
+    PolynomialDegreeIsZero,
+    /// This error occurs when the pairing check fails at being equal to the Identity point.
+    PairingCheckFailure,
```

Comment:
Since nested enums are not a thing. I guess that this is the direction. What I would suggest anyway, is to try to sepparate with comments at least the sections of errors. So that they don't get mixed up and it's clear which one should be used for each thing.

Reply:
Gotcha!